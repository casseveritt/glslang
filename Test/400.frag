#version 400 core

in vec2 c2D;
flat in int i;
out vec4 outp;
uniform sampler2D arrayedSampler[5];
uniform usampler2DRect samp2dr;
uniform isampler2DArray isamp2DA;

void main()
{
    vec4 v;
    v = texture(arrayedSampler[i], c2D);
    outp.x = gl_ClipDistance[1];

    ivec2 offsets[4];
    const ivec2 constOffsets[4] = ivec2[4](ivec2(1,2), ivec2(3,4), ivec2(15,16), ivec2(-2,0));
    uvec4 uv4 = textureGatherOffsets(samp2dr, c2D, offsets, 2);  // ERROR, offsets not constant
    uv4 = textureGatherOffsets(samp2dr, c2D, constOffsets, 2);
    vec4 v4 = textureGather(arrayedSampler[0], c2D);
    ivec4 iv4 = textureGatherOffset(isamp2DA, vec3(0.1), ivec2(1), 3);
    iv4 = textureGatherOffset(isamp2DA, vec3(0.1), ivec2(1), i);  // ERROR, last argument not const
    iv4 = textureGatherOffset(isamp2DA, vec3(0.1), ivec2(1), 4);  // ERROR, last argument out of range
    iv4 = textureGatherOffset(isamp2DA, vec3(0.1), ivec2(1), 1+2);
    iv4 = textureGatherOffset(isamp2DA, vec3(0.1), ivec2(i));

    vec4 c = gl_FragCoord;
}

layout(location = 4) in vec4 vl; // ERROR, not supported

#ifdef GL_ARB_separate_shader_objects
#extension GL_ARB_separate_shader_objects : enable
#endif

layout(location = 6) in vec4 vl2;

layout(location = 3) uniform vec3 uv3;

layout(location = 5) in vec4 gl_Color;      // ERROR, layout
noperspective in float gl_ClipDistance[4];  // ERROR, can't change qualifier

layout(origin_upper_left, pixel_center_integer) in vec4 gl_FragCoord;  // ERROR, declared after use

uniform sampler2DRectShadow u2drs;

void foo23()
{
    const ivec2[3] offsets = ivec2[3](ivec2(1,2), ivec2(3,4), ivec2(15,16));

    textureProjGradOffset(u2drs, outp, vec2(0.0), vec2(0.0), ivec2(c2D));     // ERROR, offset not constant
    textureProjGradOffset(u2drs, outp, vec2(0.0), vec2(0.0), offsets[1]);
    textureProjGradOffset(u2drs, outp, vec2(0.0), vec2(0.0), offsets[2]);     // ERROR, offset out of range
    textureProjGradOffset(u2drs, outp, vec2(0.0), vec2(0.0), ivec2(-10, 20)); // ERROR, offset out of range
}

patch in vec4 patchIn;              // ERROR
patch out vec4 patchOut;            // ERROR
