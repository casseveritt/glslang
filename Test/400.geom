#version 400 core

void main()
{
    EmitStreamVertex(1);
    EndStreamPrimitive(0);
    EmitVertex();
    EndPrimitive();
    int id = gl_InvocationID;
}

layout(invocations = 4) in outbn { int a; } bn[]; // ERROR, not on a block
layout(max_vertices = 127) out;
layout(invocations = 4) in;

#extension GL_ARB_separate_shader_objects : enable

in gl_PerVertex {      // testing input arrays with a block redeclaration, see 420.geom for without
    vec4 gl_Position;
    layout(std140, location = 3) patch float gl_PointSize; // ERRORs...
} gl_in[];

void foo()
{
    gl_in.length();  // ERROR
    gl_in[1].gl_Position;
}

in vec4 color[];
in vec4 color2[];
in vec4 colorS[3];
in vec4 colorBad[4];

void foo2()
{
    color.length(); // ERROR
    colorS.length();
}

layout(triangles) in;  // give ERROR just for colorBad

in vec4 color[3];
in vec4 color2[3];
in vec4 colorbad2[2];  // ERROR

void foo3()
{
    gl_in.length();
    color.length();
    color2.length();
    colorS.length();
}

layout(location = 4) in vec4 cva[3];
layout(location = 5) in vec4 cvb[3];
layout(location = 2) in mat3 cmc[3];  // ERROR, collision

patch in vec4 patchIn[];            // ERROR
patch out vec4 patchOut;            // ERROR

in float scalar;  // ERROR, no array

layout(max_vertices = 127, invocations = 4) out;      // ERROR
layout(invocations = 4, max_vertices = 127) in;       // ERROR
layout(max_vertices = 127, invocations = 4) uniform;  // 2 ERRORs

in inblockscalar {
    int a;
} inbls;  // ERROR, not an array

in inblocka {
    int a;
} inbla[17];  // ERROR, wrong array size
