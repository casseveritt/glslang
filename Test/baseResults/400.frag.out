400.frag
Warning, version 400 is not yet complete; some version-specific features are present, but many are missing.
ERROR: 0:18: 'textureGatherOffsets(...)' : must be a compile-time constant: offsets argument
ERROR: 0:22: 'textureGatherOffset(...)' : must be a compile-time constant: component argument
ERROR: 0:23: 'textureGatherOffset(...)' : must be 0, 1, 2, or 3: component argument
ERROR: 0:30: 'location qualifier on input' : not supported for this version or the enabled extensions 
ERROR: 0:38: 'location qualifier on uniform or buffer' : not supported for this version or the enabled extensions 
ERROR: 0:40: 'redeclaration' : cannot apply layout qualifier to gl_Color
ERROR: 0:41: 'redeclaration' : cannot change qualification of gl_ClipDistance
ERROR: 0:43: 'gl_FragCoord' : cannot redeclare after use 
ERROR: 0:51: 'texel offset' : argument must be compile-time constant 
ERROR: 0:53: 'texel offset' : value is out of range: [gl_MinProgramTexelOffset, gl_MaxProgramTexelOffset]
ERROR: 0:53: 'texel offset' : value is out of range: [gl_MinProgramTexelOffset, gl_MaxProgramTexelOffset]
ERROR: 0:54: 'texel offset' : value is out of range: [gl_MinProgramTexelOffset, gl_MaxProgramTexelOffset]
ERROR: 0:54: 'texel offset' : value is out of range: [gl_MinProgramTexelOffset, gl_MaxProgramTexelOffset]
ERROR: 0:57: 'patch' : not supported in this stage: fragment
ERROR: 0:58: 'patch' : not supported in this stage: fragment
ERROR: 15 compilation errors.  No code generated.


gl_FragCoord pixel center is integer
gl_FragCoord origin is upper left
ERROR: node is still EOpNull!
0:10  Function Definition: main( (void)
0:10    Function Parameters: 
0:?     Sequence
0:13      move second child to first child (4-component vector of float)
0:13        'v' (4-component vector of float)
0:13        Function Call: texture(s21;vf2; (4-component vector of float)
0:13          indirect index (sampler2D)
0:13            'arrayedSampler' (uniform 5-element array of sampler2D)
0:13            'i' (flat in int)
0:13          'c2D' (smooth in 2-component vector of float)
0:14      move second child to first child (float)
0:14        direct index (float)
0:14          'outp' (out 4-component vector of float)
0:14          Constant:
0:14            0 (const int)
0:14        direct index (smooth float)
0:14          'gl_ClipDistance' (smooth in unsized array of float)
0:14          Constant:
0:14            1 (const int)
0:18      Sequence
0:18        move second child to first child (4-component vector of uint)
0:18          'uv4' (4-component vector of uint)
0:18          Function Call: textureGatherOffsets(usR21;vf2;vi2[4];i1; (4-component vector of uint)
0:18            'samp2dr' (uniform usampler2DRect)
0:18            'c2D' (smooth in 2-component vector of float)
0:18            'offsets' (4-element array of 2-component vector of int)
0:18            Constant:
0:18              2 (const int)
0:19      move second child to first child (4-component vector of uint)
0:19        'uv4' (4-component vector of uint)
0:19        Function Call: textureGatherOffsets(usR21;vf2;vi2[4];i1; (4-component vector of uint)
0:19          'samp2dr' (uniform usampler2DRect)
0:19          'c2D' (smooth in 2-component vector of float)
0:19          Constant:
0:19            1 (const int)
0:19            2 (const int)
0:19            3 (const int)
0:19            4 (const int)
0:19            15 (const int)
0:19            16 (const int)
0:19            -2 (const int)
0:19            0 (const int)
0:19          Constant:
0:19            2 (const int)
0:20      Sequence
0:20        move second child to first child (4-component vector of float)
0:20          'v4' (4-component vector of float)
0:20          Function Call: textureGather(s21;vf2; (4-component vector of float)
0:20            direct index (sampler2D)
0:20              'arrayedSampler' (uniform 5-element array of sampler2D)
0:20              Constant:
0:20                0 (const int)
0:20            'c2D' (smooth in 2-component vector of float)
0:21      Sequence
0:21        move second child to first child (4-component vector of int)
0:21          'iv4' (4-component vector of int)
0:21          Function Call: textureGatherOffset(isA21;vf3;vi2;i1; (4-component vector of int)
0:21            'isamp2DA' (uniform isampler2DArray)
0:21            Constant:
0:21              0.100000
0:21              0.100000
0:21              0.100000
0:21            Constant:
0:21              1 (const int)
0:21              1 (const int)
0:21            Constant:
0:21              3 (const int)
0:22      move second child to first child (4-component vector of int)
0:22        'iv4' (4-component vector of int)
0:22        Function Call: textureGatherOffset(isA21;vf3;vi2;i1; (4-component vector of int)
0:22          'isamp2DA' (uniform isampler2DArray)
0:22          Constant:
0:22            0.100000
0:22            0.100000
0:22            0.100000
0:22          Constant:
0:22            1 (const int)
0:22            1 (const int)
0:22          'i' (flat in int)
0:23      move second child to first child (4-component vector of int)
0:23        'iv4' (4-component vector of int)
0:23        Function Call: textureGatherOffset(isA21;vf3;vi2;i1; (4-component vector of int)
0:23          'isamp2DA' (uniform isampler2DArray)
0:23          Constant:
0:23            0.100000
0:23            0.100000
0:23            0.100000
0:23          Constant:
0:23            1 (const int)
0:23            1 (const int)
0:23          Constant:
0:23            4 (const int)
0:24      move second child to first child (4-component vector of int)
0:24        'iv4' (4-component vector of int)
0:24        Function Call: textureGatherOffset(isA21;vf3;vi2;i1; (4-component vector of int)
0:24          'isamp2DA' (uniform isampler2DArray)
0:24          Constant:
0:24            0.100000
0:24            0.100000
0:24            0.100000
0:24          Constant:
0:24            1 (const int)
0:24            1 (const int)
0:24          Constant:
0:24            3 (const int)
0:25      move second child to first child (4-component vector of int)
0:25        'iv4' (4-component vector of int)
0:25        Function Call: textureGatherOffset(isA21;vf3;vi2; (4-component vector of int)
0:25          'isamp2DA' (uniform isampler2DArray)
0:25          Constant:
0:25            0.100000
0:25            0.100000
0:25            0.100000
0:25          Construct ivec2 (2-component vector of int)
0:25            'i' (flat in int)
0:27      Sequence
0:27        move second child to first child (4-component vector of float)
0:27          'c' (4-component vector of float)
0:27          'gl_FragCoord' (gl_FragCoord 4-component vector of float)
0:47  Function Definition: foo23( (void)
0:47    Function Parameters: 
0:?     Sequence
0:51      Function Call: textureProjGradOffset(sSR21;vf4;vf2;vf2;vi2; (float)
0:51        'u2drs' (uniform sampler2DRectShadow)
0:51        'outp' (out 4-component vector of float)
0:51        Constant:
0:51          0.000000
0:51          0.000000
0:51        Constant:
0:51          0.000000
0:51          0.000000
0:51        Convert float to int (2-component vector of int)
0:51          'c2D' (smooth in 2-component vector of float)
0:52      Function Call: textureProjGradOffset(sSR21;vf4;vf2;vf2;vi2; (float)
0:52        'u2drs' (uniform sampler2DRectShadow)
0:52        'outp' (out 4-component vector of float)
0:52        Constant:
0:52          0.000000
0:52          0.000000
0:52        Constant:
0:52          0.000000
0:52          0.000000
0:52        Constant:
0:52          3 (const int)
0:52          4 (const int)
0:53      Function Call: textureProjGradOffset(sSR21;vf4;vf2;vf2;vi2; (float)
0:53        'u2drs' (uniform sampler2DRectShadow)
0:53        'outp' (out 4-component vector of float)
0:53        Constant:
0:53          0.000000
0:53          0.000000
0:53        Constant:
0:53          0.000000
0:53          0.000000
0:53        Constant:
0:53          15 (const int)
0:53          16 (const int)
0:54      Function Call: textureProjGradOffset(sSR21;vf4;vf2;vf2;vi2; (float)
0:54        'u2drs' (uniform sampler2DRectShadow)
0:54        'outp' (out 4-component vector of float)
0:54        Constant:
0:54          0.000000
0:54          0.000000
0:54        Constant:
0:54          0.000000
0:54          0.000000
0:54        Constant:
0:54          -10 (const int)
0:54          20 (const int)
0:?   Linker Objects
0:?     'c2D' (smooth in 2-component vector of float)
0:?     'i' (flat in int)
0:?     'outp' (out 4-component vector of float)
0:?     'arrayedSampler' (uniform 5-element array of sampler2D)
0:?     'samp2dr' (uniform usampler2DRect)
0:?     'isamp2DA' (uniform isampler2DArray)
0:?     'gl_ClipDistance' (smooth in 4-element array of float)
0:?     'vl' (layout(location=4 ) smooth in 4-component vector of float)
0:?     'vl2' (layout(location=6 ) smooth in 4-component vector of float)
0:?     'uv3' (layout(location=3 ) uniform 3-component vector of float)
0:?     '__anon__0' (in block{in float gl_FogFragCoord, in unsized array of 4-component vector of float gl_TexCoord, smooth in 4-component vector of float gl_Color, in 4-component vector of float gl_SecondaryColor})
0:?     '__anon__0' (in block{in float gl_FogFragCoord, in unsized array of 4-component vector of float gl_TexCoord, smooth in 4-component vector of float gl_Color, in 4-component vector of float gl_SecondaryColor})
0:?     'gl_FragCoord' (gl_FragCoord 4-component vector of float)
0:?     'gl_FragCoord' (gl_FragCoord 4-component vector of float)
0:?     'u2drs' (uniform sampler2DRectShadow)
0:?     'patchIn' (smooth patch in 4-component vector of float)
0:?     'patchOut' (patch out 4-component vector of float)


Linked fragment stage:


gl_FragCoord pixel center is integer
gl_FragCoord origin is upper left

